lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -0,0 +1,372 @@
+
+        self._clean_up_credentials()
+
+    def _clean_up_credentials(self) -> None:
+        """Clean up roles and policies that are no longer needed by autounseal.
+
+        This method will remove any roles and policies that are no longer
+        used by any of the existing relations. It will also detect any orphaned
+        keys (keys that are not associated with any relation) and log a warning.
+        """
+        self._clean_up_roles()
+        self._clean_up_policies()
+        self._detect_orphaned_keys()
+
+    def _detect_orphaned_keys(self) -> None:
+        existing_keys = self._get_existing_keys()
+        relation_key_names = [
+            VaultAutounsealRelationDetails(relation).key_name
+            for relation in self._juju_facade.get_active_relations(self._provides.relation_name)
+        ]
+        orphaned_keys = [key for key in existing_keys if key not in relation_key_names]
+        if not orphaned_keys:
+            return
+        logging.warning(
+            f"Orphaned autounseal keys were detected: {orphaned_keys}. If you are sure these are no longer needed, you may manually delete them using the vault CLI to suppress this message."
+            " To delete a key, use the command `vault delete charm-autounseal/keys/<key_name>`."
+        )
+        for key_name in orphaned_keys:
+            deletion_allowed = self._detect_if_deletion_allowed(key_name)
+            if not deletion_allowed:
+                self._allow_key_deletion(key_name)
+
+    def _allow_key_deletion(self, key_name) -> None:
+        logger.info("Allowing deletion of key %s", key_name)
+        self._client.write(f"{self.mount_path}/keys/{key_name}/config", {"deletion_allowed": True})
+
+    def _detect_if_deletion_allowed(self, key_name) -> bool:
+        data = self._client.read(f"{self.mount_path}/keys/{key_name}")
+        return data["deletion_allowed"]
+
+    def _clean_up_roles(self) -> None:
+        existing_roles = self._get_existing_roles()
+        relation_role_names = [
+            VaultAutounsealRelationDetails(relation).approle_name
+            for relation in self._juju_facade.get_active_relations(self._provides.relation_name)
+        ]
+        for role in existing_roles:
+            if role not in relation_role_names:
+                logging.info("Removing unused role: %s", role)
+                self._client.delete_role(role)
+
+    def _clean_up_policies(self) -> None:
+        existing_policies = self._get_existing_policies()
+        relation_policy_names = [
+            VaultAutounsealRelationDetails(relation).policy_name
+            for relation in self._juju_facade.get_active_relations(self._provides.relation_name)
+        ]
+        for policy in existing_policies:
+            if policy not in relation_policy_names:
+                logging.info("Removing unused policy: %s", policy)
+                self._client.delete_policy(policy)
+
+    def _create_key(self, key_name) -> None:
```

Comment:
We're missing the input type

Reply:
https://github.com/canonical/vault-k8s-operator/pull/534/commits/ab549ee6c4957d3caa79a70f47db804a4e9c142b